accessor.go

```diff
@@ -95,3 +112,28 @@ func (sa *ShardAccessor) Close() error {
 	tsk := &task{op: OpShardRelease, shard: sa.shard}
 	return sa.shard.d.queueTask(tsk, sa.shard.d.externalCh)
 }
+
+// readerAtWrapper is a wrapper around an io.ReaderAt that implements io.Reader.
+type readerAtWrapper struct {
+	readerAt   io.ReaderAt
+	readOffset int64
+}
+
+func (w *readerAtWrapper) Read(p []byte) (n int, err error) {
+	if w.readerAt == nil {
+		return 0, errors.New("readerAtWrapper: readerAt is nil")
+	}
+
+	n, err = w.readerAt.ReadAt(p, w.readOffset)
+	w.readOffset += int64(n)
+
+	if err == io.EOF {
+		return n, err
+	}
+
+	if err != nil {
+		return n, errors.New("readerAtWrapper: error reading from the underlying ReaderAt")
```

Comment:
return the actual err as well. use fmt.Errorf

Reply:
done